data/heuristics.go

```diff
@@ -0,0 +1,97 @@
+package data
+
+import "regexp"
+
+type (
+	Heuristics []Matcher
+
+	Matcher interface {
+		Match(data []byte) bool
+	}
+
+	Languages struct {
+		langs []string
+	}
+
+	Rule interface {
+		GetLanguages() []string
+	}
+)
+
+func (h *Heuristics) Match(data []byte) []string {
+	var matchedLangs []string
+	for _, matcher := range *h {
+		if matcher.Match(data) {
+			for _, langOrAlias := range matcher.(Rule).GetLanguages() {
+				lang, ok := LanguageByAlias(langOrAlias)
+				if !ok { // should never happen
+					// language name/alias in heuristics.yml is not consistent with languages.yml
+					// but we do not surface any error on the API
+					continue
+				}
+				matchedLangs = append(matchedLangs, lang)
+			}
+			break
+		}
+	}
+	return matchedLangs
+}
+
+// matchString is a convenience used only in tests.
+func (h *Heuristics) matchString(data string) []string {
+	return h.Match([]byte(data))
+}
+
+// GetLanguages returns languages, defined by this data.Rule.
+func (l *Languages) GetLanguages() []string {
+	return l.langs
+}
+
+type OrRule struct {
```

Comment:
Can we drop the `Rule` suffixes from these? If there are conflicts, they probably deserve a package of their own.

Reply:
Those are rules for auto-generated rule-based matching system, so as a reader I found it convenient to have a common suffix to remind about that.

Can drop the suffix ofc, but I'm not sure how/what does it help with.

Reply:
Ah, I didn't realize this was machine generated. Can you make the generator insert a comment at the top of the file along the lines of `// This file was generated, do not edit.` ?

Reply:
As far as the names: It's generally worthwhile to avoid redundant information in names. If all or many of the names in a given package share a common prefix or suffix, that is often a sign that either (a) it should be implicit, or (b) there is a sub-package that can be pulled out.

It's definitely not a blocker by any means, but it does affect the API of the code in use: "AndRule" and "OrRule" are kind of unnatural for someone writing an expression as opposed to simply "And" and "Or". Perhaps if this were package `rule`, we could have `rule.And` and `rule.Or` to get the same effect without decorating the names.

That said: This is not urgent, and even if we decide to do so it can wait.

Reply:
> Ah, I didn't realize this was machine generated. Can you make the generator insert a comment at the top of the file along the lines of `// This file was generated, do not edit.` ?

sorry for confusion - this code is human-written, but is only used inside machine-generated datastructures in `content.go`

Reply:
addressed in f299f2050994c8a0be3f197f53707196e761eda6